File: src/Scheduler.java
import java.util.PriorityQueue;

/**
 * Scheduler.java — Ordonnanceur d'événements (liste chronologique).
 *
 * Choix d'implémentation : PriorityQueue (tas binaire min-heap de Java).
 *   - addEvent  : insertion en O(log n) — le tas se réorganise seul, sans sort().
 *   - getEvent  : extraction du minimum (= l'événement le plus proche) en O(log n).
 * C'est exactement ce que le sujet demande : "insertion directe au bon endroit,
 * interdit d'appeler sort() à chaque insertion".
 *
 * L'ordre est défini par Event.compareTo() : chronologique, puis par eventID.
 */
public class Scheduler {

    // Tas binaire : le premier élément est toujours l'événement le plus tôt
    private PriorityQueue<Event> tas;
    private double tempsCourant;  // temps du dernier événement extrait

    public Scheduler() {
        this.tas          = new PriorityQueue<>();
        this.tempsCourant = 0.0;
    }

    /**
     * Insère un événement au bon endroit dans le tas (O(log n)).
     * La PriorityQueue garantit l'ordre — pas de sort() appelé.
     */
    public void addEvent(Event e) {
        tas.offer(e);
    }

    /**
     * Extrait et retourne le prochain événement (le plus tôt).
     * Met à jour le temps courant.
     * Retourne null si la liste est vide.
     */
    public Event getEvent() {
        Event e = tas.poll();
        if (e != null) {
            tempsCourant = e.getEventTime();
        }
        return e;
    }

    /** Retourne le temps courant (celui du dernier événement extrait). */
    public double getCurrentTime() {
        return tempsCourant;
    }

    /** Indique s'il reste des événements à traiter. */
    public boolean hasEvents() {
        return !tas.isEmpty();
    }
}

File: python/src/scheduler.py
"""
scheduler.py - Classe Scheduler : l'ordonnanceur d'événements.

Le scheduler maintient les événements futurs en ordre chronologique.
Le sujet interdit d'appeler sort() à chaque insertion. On utilise un tas
binaire (heapq) : AddEvent insère en O(log n) et GetEvent extrait le
minimum (le prochain événement) en O(log n), sans jamais re-trier la liste
entière. L'ordre chronologique est donc garanti par construction.
"""

import heapq


class Scheduler:
    def __init__(self):
        self._tas = []            # tas binaire d'objets Event
        self._temps_courant = 0.0  # temps de l'événement en cours

    def AddEvent(self, e):
        """Insère un événement au bon endroit (ordre chronologique maintenu)."""
        heapq.heappush(self._tas, e)

    def GetEvent(self):
        """Extrait et retourne le prochain événement (le plus tôt)."""
        e = heapq.heappop(self._tas)
        self._temps_courant = e.GetEventTime()
        return e

    def GetCurrentTime(self):
        """Retourne le temps courant (celui du dernier événement extrait)."""
        return self._temps_courant

    def HasEvents(self):
        """Indique s'il reste des événements à traiter."""
        return len(self._tas) > 0
